onnx/checker.h

```diff
@@ -142,8 +142,8 @@ void check_model_local_functions(
     const CheckerContext& ctx,
     const LexicalScopeContext& parent_lex);
 
-void check_model(const ModelProto& model);
-void check_model(const std::string& model_path);
+void check_model(ModelProto& model, bool full_check = false);
```

Comment:
`void check_model(const ModelProto& model)` -> `void check_model(ModelProto& model)` brings a breaking change for C++ users and also unnecessarily refuses const models when full checking is not needed. How about 

```suggestion
void check_model(const ModelProto& model);
void check_model(ModelProto& model, bool full_check);
```

Reply:
Thanks for the suggestion, not breaking the API is better in any case. I added new commits addressing the comments.

Reply:
I have a concern: if users call checker with full_check, did they expect the ONNX model will get updated (shape_inference will bring inferred shapes in value_info)?

Current behavior:
- check_model+full_check with model path (Python and C++ API): the original model will be overrode by the inferred model thru model path (it's by default from onnx.shape_inference_path)
- check_model+full_check with ModelProto (Python and C++ API): the original model won't be updated

After this PR:
- check_model+full_check with model path (Python and C++ API): same as current behavior
- check_model+full_check with ModelProto (Python API): the original model won't be updated
- check_model+full_check with ModelProto (C++ API): the original ModelProto will be overrode by the inferred model

IMO, ideally all APIs behaviors for check_model should be consistent to prevent confusion. It seems weird to me that the ONNX model will be updated after onnx.checker with full_check. Still, it's an existing issue (check_model does not update the model, but check_model_path does) so it should not be the blocker to this PR. If both of you also think it's a valid concern, probably we can improve it in another PR.

Reply:
Yes, it would be helpful to ensure a consistent and orthogonal API. Do we need full-check and const-ness of the model interact or can they be orthogonal/independent? It would also be a good idea to add some comment/documentation for all of this.